receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go

```diff
@@ -0,0 +1,402 @@
+	GetContainerInfo(deviceID string, resourceName string) *stores.ContainerInfo
+}
+
+type efaStore struct {
+	timestamp time.Time
+	devices   *efaDevices
+}
+
+// efaDevices is a collection of every Amazon Elastic Fabric Adapter (EFA) device in
+// /sys/class/infiniband.
+type efaDevices map[efaDeviceName]*efaCounters
+
+type efaDeviceName string
+
+// efaCounters contains counter values from files in
+// /sys/class/infiniband/<Name>/ports/<Port>/hw_counters
+// for a single port of one Amazon Elastic Fabric Adapter device.
+type efaCounters struct {
+	rdmaReadBytes      uint64 // hw_counters/rdma_read_bytes
+	rdmaWriteBytes     uint64 // hw_counters/rdma_write_bytes
+	rdmaWriteRecvBytes uint64 // hw_counters/rdma_write_recv_bytes
+	rxBytes            uint64 // hw_counters/rx_bytes
+	rxDrops            uint64 // hw_counters/rx_drops
+	txBytes            uint64 // hw_counters/tx_bytes
+}
+
+func NewEfaSyfsScraper(logger *zap.Logger, decorator stores.Decorator, podResourcesStore podResourcesStore) *Scraper {
+	ctx, cancel := context.WithCancel(context.Background())
+	podResourcesStore.AddResourceName(efaK8sResourceName)
+	e := &Scraper{
+		collectionInterval: defaultCollectionInterval,
+		cancel:             cancel,
+		sysFsReader:        defaultSysFsReader,
+		deltaCalculator:    metrics.NewMetricCalculator(calculateDelta),
+		decorator:          decorator,
+		podResourcesStore:  podResourcesStore,
+		store:              new(efaStore),
+		logger:             logger,
+	}
+
+	go e.startScrape(ctx)
+
+	return e
+}
+
+func calculateDelta(prev *metrics.MetricValue, val any, _ time.Time) (any, bool) {
+	var deltaValue uint64
+	if prev != nil {
+		deltaValue = val.(uint64) - prev.RawValue.(uint64)
+	} else {
+		return deltaValue, false
+	}
+	return deltaValue, true
+}
+
+func (s *Scraper) Shutdown() {
+	if s.cancel != nil {
+		s.cancel()
+	}
+}
+
+func (s *Scraper) GetMetrics() []pmetric.Metrics {
+	var result []pmetric.Metrics
+
+	store := s.store
+	for deviceName, counters := range *store.devices {
+		containerInfo := s.podResourcesStore.GetContainerInfo(string(deviceName), efaK8sResourceName)
+		if containerInfo == nil {
+			continue
+		}
+
+		containerMetric := stores.NewCIMetric(ci.TypeContainerEFA, s.logger)
+		podMetric := stores.NewCIMetric(ci.TypePodEFA, s.logger)
+		nodeMetric := stores.NewCIMetric(ci.TypeNodeEFA, s.logger)
+
+		s.fillMetric(containerMetric, ci.TypeContainerEFA, containerInfo.Namespace, containerInfo.PodName,
+			containerInfo.ContainerName, string(deviceName), store.timestamp, counters)
+		s.fillMetric(podMetric, ci.TypePodEFA, containerInfo.Namespace, containerInfo.PodName,
+			containerInfo.ContainerName, string(deviceName), store.timestamp, counters)
+		s.fillMetric(nodeMetric, ci.TypeNodeEFA, containerInfo.Namespace, containerInfo.PodName,
+			containerInfo.ContainerName, string(deviceName), store.timestamp, counters)
+
+		for _, m := range []stores.CIMetric{containerMetric, podMetric, nodeMetric} {
```

Comment:
you can merge this for loop into the loop where it calls `decorator.Decorate`. It's technically doing partial decoration on all metric types.

Reply:
I like to keep it separate so the series of steps is clear: 1) populate metric values, 2) add tags that the EFA scraper is responsible for, 3) pass to the decorator to fill in other metadata, and 4) convert to OTLP.

Reply:
makes sense 👍